File: src/main/java/dev/esophose/discordbot/listener/Listener.java
package dev.esophose.discordbot.listener;

import discord4j.core.event.domain.Event;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public abstract class Listener<T extends Event> {

    private final List<Class<T>> eventClasses;

    @SuppressWarnings("unchecked")
    public Listener(Class<?>... eventClasses) {
        this.eventClasses = Arrays.stream(eventClasses).map(x -> (Class<T>) x).collect(Collectors.toList());
    }

    public List<Class<T>> getEventClasses() {
        return this.eventClasses;
    }

    public abstract void execute(T event);

}

File: src/main/kotlin/dev/esophose/discordbot/listener/Listener.kt
package dev.esophose.discordbot.listener

import discord4j.core.event.domain.Event
import java.util.Arrays
import java.util.stream.Collectors
import kotlin.reflect.KClass

abstract class Listener<T : Event>(vararg eventClasses: KClass<*>) {

    val eventClasses: List<KClass<T>>

    init {
        @Suppress("UNCHECKED_CAST")
        this.eventClasses = Arrays.stream(eventClasses).map { it as KClass<T> }.collect(Collectors.toList())
    }

    abstract fun execute(event: T)

}
